File: src/token_goat/languages/ini_idx.py
"""INI / CFG / .env extractor — one Section per ``[section]`` header.

INI-family configuration files are line-oriented and unambiguous: a
``[name]`` header at column 0 opens a section that spans every following
line until the next header or EOF.  ``.env`` (dotenv) files have no section
syntax at all — they are flat ``KEY=value`` pairs — so for those we emit
one ``env_key`` symbol per top-level assignment and skip sections entirely.

Why a custom scanner rather than :mod:`configparser`:

* :mod:`configparser` parses to a dict and discards source positions.  We
  need start/end line numbers so ``token-goat section`` can slice the source
  file back out.

* INI dialects vary (Windows ``;`` comments vs Unix ``#``; multi-line values
  with continuation indent; spaces in keys).  A targeted line scanner gives
  predictable, low-surprise behaviour without inheriting configparser's
  strictness on edge cases that token-goat does not need to enforce.

Section model
-------------
* ``heading``: the bracketed name, lowercased and trimmed.  Dotted/colon-
  separated sections like ``[tool.black]`` or ``[mysqld:replica]`` are kept
  verbatim so callers can target the exact name they see in the file.
* ``level``: always 1 — INI has no nested headers.
* ``line``: 1-based line of the header.
* ``end_line``: 1-based last line of the section's content (the line
  immediately before the next header, or EOF for the trailing entry).

The ``.env`` path emits no sections — only the per-key symbols — because
treating each top-level key as a "section" would produce one entry per
line and inflate the index for what is already a small flat file.
"""
from __future__ import annotations

__all__ = ["extract", "extract_env"]

import re

from ..parser import ImpExp, Ref, Section, Symbol
from ..util import get_logger
from . import common

_LOG = get_logger("languages.ini_idx")

# Column-0-anchored ``[name]`` header.  We allow letters, digits, underscores,
# hyphens, dots, colons, and slashes in the name — this covers every dialect
# I've seen in the wild (``[tool.black]`` in setup.cfg, ``[mysqld:replica]``
# in my.cnf, ``[group/sub]`` in PHP-FPM pools) without admitting whitespace
# or quotes that would indicate a malformed line.
_HEADER_RE = re.compile(r"^\[([A-Za-z0-9_\-.:/]+)\]\s*(?:[;#].*)?$")

# Maximum number of headers indexed per file.  Real INI files top out in the
# low tens; the cap is generous so a hand-typed config never hits it but
# tight enough to bound a pathological generated file (Apache ``vhost`` dumps,
# Windows ``.ini`` exports with thousands of entries).
_MAX_SECTIONS: int = 200
# Maximum length of a section header we accept.  Real names are short.
_MAX_HEADING_LEN: int = 200


def extract(
    source: bytes, rel_path: str
) -> tuple[list[Symbol], list[Ref], list[ImpExp], list[Section]]:
    """Extract INI/CFG ``[section]`` headers as Section + Symbol entries.

    Refs and imports are always empty for INI files — there is no cross-file
    reference model in this format.
    """
    text = common.decode_source_text(source, _LOG, "ini_idx")
    if text is None:
        return [], [], [], []

    lines = text.split("\n")
    sections: list[Section] = []
    symbols: list[Symbol] = []

    for idx, line in enumerate(lines, start=1):
        # Strip a UTF-8 BOM if present at file start so the column-0 anchor
        # still matches a header on line 1 of a BOM-saved file (Notepad on
        # Windows defaults to UTF-8 with BOM for plain-text saves).
        candidate = common.bom_strip_first_line(line, idx)
        if not candidate or candidate[0] != "[":
            continue
        m = _HEADER_RE.match(candidate)
        if m is None:
            continue
        name = m.group(1).strip()
        if not name or len(name) > _MAX_HEADING_LEN:
            continue
        sections.append(Section(heading=name, level=1, line=idx))
        symbols.append(Symbol(name=name, kind="ini_section", line=idx))
        if len(sections) >= _MAX_SECTIONS:
            break

    # End-line computation: each section spans from its header through the
    # line before the next header (or EOF for the trailing section).  This is
    # the same shape as TOML — both formats are flat at the source level even
    # when their names look hierarchical.
    common.assign_flat_end_lines(sections, len(lines))

    return symbols, [], [], sections


# A flat ``KEY=value`` assignment at column 0.  ``=`` and ``:`` are both
# accepted as the separator because real-world ``.env`` and ``.envrc`` files
# use either; the key body matches the standard shell-identifier character
# class.  Lines with leading whitespace are intentionally skipped — they are
# either continuation values or invalid — and lines starting with ``#`` /
# ``;`` are comments.
_ENV_KEY_RE = re.compile(r"^([A-Za-z_][A-Za-z0-9_]*)\s*[:=]")

# Maximum number of env keys captured per file.  Production ``.env`` files
# rarely exceed a few dozen; the cap is conservative against pathological
# auto-generated dumps.
_MAX_ENV_KEYS: int = 200


def extract_env(
    source: bytes, rel_path: str
) -> tuple[list[Symbol], list[Ref], list[ImpExp], list[Section]]:
    """Extract ``.env`` / ``.envrc`` top-level keys as ``env_key`` symbols.

    Sections, refs, and imports are always empty for dotenv files: the format
    is flat by design and there is no surrounding "block" to slice.  Each
    captured key carries its 1-based line number so ``token-goat symbol``
    points at the assignment.
    """
    text = common.decode_source_text(source, _LOG, "ini_idx")
    if text is None:
        return [], [], [], []

    symbols: list[Symbol] = []
    for idx, line in enumerate(text.split("\n"), start=1):
        candidate = common.bom_strip_first_line(line, idx)
        if not candidate or candidate[0] in "#;":
            continue
        # Reject leading whitespace defensively: continuation lines and shell
        # heredoc bodies must not be mistaken for key assignments.
        if candidate[0] in " \t":
            continue
        m = _ENV_KEY_RE.match(candidate)
        if m is None:
            continue
        name = m.group(1)
        if not name or len(name) > _MAX_HEADING_LEN:
            continue
        symbols.append(Symbol(name=name, kind="env_key", line=idx))
        if len(symbols) >= _MAX_ENV_KEYS:
            break

    return symbols, [], [], []

File: src/languages/ini_idx.ts
/**
 * INI/CFG section extractor and .env key extractor.
 *
 * `extractIni` — one symbol+section per `[header]` at column 0.
 * `extractEnv` — one symbol per `KEY=value` assignment at column 0, with an optional leading `export `.
 */

import type { SymbolEntry } from '../parser_types.js'
import type { MiniSection } from './common.js'
import { assignFlatEndLines, propagateEndLinesToSymbols, makeLineSymbol } from './common.js'

const MAX_SECTIONS = 200
const MAX_HEADING_LEN = 200
const MAX_ENV_KEYS = 200

// Column-0-anchored [name] header. Captures anything up to the next `]` — covers plain names like [tool.black]/[mysqld:replica]/[group/sub] as well as quoted/spaced git-config-style subsection headers like [branch "master"], which a name-charset allowlist would reject.
const HEADER_RE = /^\[([^\]\r\n]+)\]\s*(?:[;#].*)?$/

// Optional leading `export ` (shell-sourced .env / direnv .envrc) is consumed so the captured key is the variable name, not the literal `export`. A var literally named `export` (`export=5`, no following space) still captures as `export` because the prefix group requires whitespace. A bare URL on its own line (e.g. `https://example.com`) must NOT match as a false `https` key - the colon there is a URL scheme separator immediately followed by `//`, not a key/value split. Mirrors the same `:(?!\/\/)` guard the live section reader's KEYVALUE_HEADER_RE already applies (section_reader.ts). Matches ENV_KEYVALUE_HEADER_RE in section_reader.ts - includes '.'/'-' so keys like NODE-ENV or DB.HOST are indexed the same as the live section reader finds them.
const ENV_KEY_RE = /^(?:export\s+)?([A-Za-z_][\w.-]*)\s*(?:=|:(?!\/\/))/

export function extractIni(content: string, filePath: string): SymbolEntry[] {
  const symbols: SymbolEntry[] = []
  const sections: MiniSection[] = []
  const seen = new Set<string>()
  const lines = content.split(/\r?\n/)
  const totalLines = lines.length

  for (let i = 0; i < lines.length; i++) {
    if (symbols.length >= MAX_SECTIONS) break
    const line = lines[i] ?? ''
    if (!line || line[0] !== '[') continue
    const m = HEADER_RE.exec(line)
    if (m === null) continue
    const name = m[1]?.trim() ?? ''
    if (!name || name.length > MAX_HEADING_LEN) continue
    const key = `${name}\0${i + 1}`
    if (seen.has(key)) continue
    seen.add(key)
    symbols.push(makeLineSymbol(filePath, name, 'ini_section', i + 1))
    sections.push({ heading: name, level: 1, line: i + 1, endLine: i + 1 })
  }

  sections.sort((a, b) => a.line - b.line)
  assignFlatEndLines(sections, totalLines)

  return propagateEndLinesToSymbols(symbols, sections)
}

/** Returns true if the quote char `q` at index `i` in `line` is escaped: an odd count of
 * consecutive backslashes immediately precedes it (each `\\` pair is one literal backslash, so
 * only an odd run actually escapes the quote — an escaped backslash before the quote must not
 * count as escaping it). Single quotes have no escape semantics in dotenv/POSIX, so this check
 * is skipped entirely for them — any `'` closes the value regardless of what precedes it. */
function _isEscapedQuote(line: string, i: number, q: string): boolean {
  if (q === "'") return false
  let backslashes = 0
  let j = i - 1
  while (j >= 0 && line[j] === '\\') {
    backslashes++
    j--
  }
  return backslashes % 2 === 1
}

/** Returns true if `line` contains an unescaped occurrence of the open quote char `q`, closing it. */
export function _lineClosesQuote(line: string, q: string): boolean {
  for (let i = 0; i < line.length; i++) {
    if (line[i] === q && !_isEscapedQuote(line, i, q)) return true
  }
  return false
}

/** If a value's leading char is a quote that isn't closed again later on the same line, returns
 * that quote char (the value continues, embedded-newline-style, onto following lines); else null. */
export function _detectOpenQuote(value: string): string | null {
  const trimmed = value.replace(/^\s+/, '')
  const q = trimmed[0]
  if (q !== '"' && q !== "'") return null
  for (let i = 1; i < trimmed.length; i++) {
    if (trimmed[i] === q && !_isEscapedQuote(trimmed, i, q)) return null
  }
  return q
}

export function extractEnv(content: string, filePath: string): SymbolEntry[] {
  const symbols: SymbolEntry[] = []
  const seen = new Set<string>()
  const lines = content.split(/\r?\n/)
  // Tracks a quote char opened on a prior line whose multi-line value hasn't closed yet, so a continuation line's content (which can look like its own `KEY=value` assignment) is never re-scanned as a new key.
  let openQuote: string | null = null

  for (let i = 0; i < lines.length; i++) {
    if (symbols.length >= MAX_ENV_KEYS) break
    const line = lines[i] ?? ''

    if (openQuote !== null) {
      if (_lineClosesQuote(line, openQuote)) openQuote = null
      continue
    }

    if (!line || line[0] === '#' || line[0] === ';' || line[0] === ' ' || line[0] === '\t') continue
    const m = ENV_KEY_RE.exec(line)
    if (m === null) continue
    const name = m[1]?.trim() ?? ''
    if (!name || name.length > MAX_HEADING_LEN) continue
    const key = `${name}\0${i + 1}`
    if (seen.has(key)) continue
    seen.add(key)
    symbols.push(makeLineSymbol(filePath, name, 'env_key', i + 1))
    openQuote = _detectOpenQuote(line.slice(m[0].length))
  }

  return symbols
}
